src/octopath.cpp

```diff
@@ -4,92 +4,139 @@
+      algebra_3d::Line3d line(p1, p2);
+      line.DistancePoint2Line(p, &dist);
+
+      // Check for collision
+      if (free_space_only) {
+        col_check = CheckCollision(p1, p2);
+      } else {
+        col_check = CheckOccupancy(p1, p2);
+      }
+
+      if (col_check == 1) {
+        continue;
+      }
+
+      // If doesn't collide, check if new nodes get closer to walls
+      // We don't allow to remove waypoints that decrease the distance to walls
+      std::vector<octomap::point3d> p1_to_p2_nodes, p1_to_p_to_p2_nodes;
+      this->GetNodesBetweenWaypoints(p1, p2, add_final_waypoint, &p1_to_p2_nodes);
+      this->GetNodesBetweenWaypoints(p1, p, do_not_add_final_waypoint, &p1_to_p_to_p2_nodes);
+      this->GetNodesBetweenWaypoints(p, p2, add_final_waypoint, &p1_to_p_to_p2_nodes);
+      if (this->AverageObstaclePathCost(p1_to_p2_nodes) > this->AverageObstaclePathCost(p1_to_p_to_p2_nodes)) {
+        continue;
+      }
+
+      if (dist > max_dist) {
+        max_dist = dist;
+        delete_index = i;
+      }
     }
-
-    // Compress the remaining points
-    static double max_dist;
-    int col_check;
-    while (compressed_points > min_points) {
-        // first find the point that deviates the least in the whole set
-        max_dist = -1;
-        delete_index = -1;
-        for (int i = 1; i < compressed_points-1; i++) {
-            p1 << (*compressed_path)[i-1][0],
-                  (*compressed_path)[i-1][1],
-                  (*compressed_path)[i-1][2];
-            p2 << (*compressed_path)[i+1][0],
-                  (*compressed_path)[i+1][1],
-                  (*compressed_path)[i+1][2];
-            p  << (*compressed_path)[i][0],
-                  (*compressed_path)[i][1],
-                  (*compressed_path)[i][2];
-            algebra_3d::Line3d line(p1, p2);
-            line.DistancePoint2Line(p, &dist);
-
-            // Check for collision
-            if (free_space_only) {
-                col_check = CheckCollision(p1, p2);
-            } else {
-                col_check = CheckOccupancy(p1, p2);
-            }
-            if ((dist > max_dist) && (col_check != 1)) {
-                max_dist = dist;
-                delete_index = i;
-            }
-        }
-        if (delete_index > 0) {
-            compressed_path->erase(compressed_path->begin() + delete_index);
-            compressed_points = compressed_points - 1;
-        } else {
-            break;
-        }
+    if (delete_index > 0) {
+      compressed_path->erase(compressed_path->begin() + delete_index);
+      compressed_points = compressed_points - 1;
+    } else {
+      break;
     }
+  }
+
+  // ROS_INFO("Compressed points: %d", compressed_points);
+}
 
-    // ROS_INFO("Compressed points: %d", compressed_points);
+// Function that computes an added cost to traverse a node based on its
+// distance to the nearest obstacle
+double OctoClass::NearestObstaclePathCost(const octomap::point3d &node_position) {
```

Comment:
This function is used in A* to add an extra cost to traverse waypoints that are close to walls (like in a cost map)